[enh] add offline engine for sqlite database

return42 opened the pull request:
To test & demonstrate this implementation download:

  https://liste.mediathekview.de/filmliste-v2.db.bz2

and unpack into searx/data/filmliste-v2.db, in your settings.yml define a sqlite engine named "demo"::

    - name : demo
      engine : sqlite
      shortcut: demo
      categories: general
      result_template: default.html
      database : searx/data/filmliste-v2.db
      query_str :  >-
        SELECT title || ' (' || time(duration, 'unixepoch') || ')' AS title,
               COALESCE( NULLIF(url_video_hd,''), NULLIF(url_video_sd,''), url_video) AS url,
               description AS content
          FROM film
         WHERE title LIKE :wildcard OR description LIKE :wildcard
         ORDER BY duration DESC
      disabled : False

Query to test: `!demo concert`

This is a rewrite of the implementation from commit [1]

[1] https://github.com/searx/searx/commit/8e90a214ce5cf60e0640e2ba812fa432d7bdef2f

Suggested-by: @virtadpt https://github.com/searx/searx/issues/2808

dalf commented on searx/engines/sqlite.py:
```diff
@@ -0,0 +1,49 @@
+    cur.execute(query_to_run, query_params)
+    results = _fetch_results(cur)
+    cur.close()
+    connection.close()
```
To avoid memory / connection / cursor leak (??) :
```python
connection = sqlite3.connect(database)
try:
  cur = connection.cursor()
  try:
    cur.execute(query_to_run, query_params)
    return _fetch_results(cur)
  finally:
    cur.close()
finally:
  connection.close()
```

(not tested)

return42 replied:
> To avoid memory / connection / cursor leak (??) :

I will implement a [context manager](https://docs.python.org/3/library/stdtypes.html#context-manager-types).

dalf replied:
I haven't seen a context manager for sqlite except for SQL transaction:
https://docs.python.org/3/library/sqlite3.html#using-the-connection-as-a-context-manager

return42 replied:
> I haven't seen a context manager for sqlite

I have already implemented one .. wait until I have rebased this PR ...

dalf commented on searx/engines/sqlite.py:
```diff
@@ -0,0 +1,49 @@
+
+def search(query, params):
+    query_params = {'query': query}
+    query_to_run = query_str + ' LIMIT {0} OFFSET {1}'.format(limit, (params['pageno'] - 1) * limit)
```
To avoid any `.format` (??):
```python
query_params = {'query': query, 'limit': limit, 'offset': (params['pageno'] - 1) * limit}
query_to_run = query_str + ' LIMIT :limit OFFSET :offset'
```

(not tested)

return42 replied:
> To avoid any .format (??)

LGTM, thanks! / BTW: using _execute parameters_ `cur.execute(..., query_params)` is more safe from SQL injection!

return42 committed:
[enh] add offline engine for sqlite database

To test & demonstrate this implementation download:

  https://liste.mediathekview.de/filmliste-v2.db.bz2

and unpack into searx/data/filmliste-v2.db, in your settings.yml define a sqlite
engine named "demo"::

    - name : demo
      engine : sqlite
      shortcut: demo
      categories: general
      result_template: default.html
      database : searx/data/filmliste-v2.db
      query_str :  >-
        SELECT title || ' (' || time(duration, 'unixepoch') || ')' AS title,
               COALESCE( NULLIF(url_video_hd,''), NULLIF(url_video_sd,''), url_video) AS url,
               description AS content
          FROM film
         WHERE title LIKE :wildcard OR description LIKE :wildcard
         ORDER BY duration DESC
      disabled : False

Query to test: "!demo concert"

This is a rewrite of the implementation from commit [1]

[1] https://github.com/searx/searx/commit/8e90a214ce5cf60e0640e2ba812fa432d7bdef2f

Suggested-by: @virtadpt https://github.com/searx/searx/issues/2808
Signed-off-by: Markus Heiser <markus@darmarit.de>

dalf commented on searx/engines/sqlite.py:
```diff
@@ -0,0 +1,78 @@
+paging = True
+result_template = 'key-value.html'
+
+class SQLiteDB:
```
IMO the function version is more readable:
```python
from contextlib import contextmanager

@contextmanager
def sqlite_cursor(database_url):
	try:
		connect = sqlite3.connect(database_url)
		connect.row_factory = sqlite3.Row
		cursor = self.connect.cursor()
		try:
			yield cursor
		finally:
			cursor.close()
	finally:
		if connect:
			connect.close()
```

return42 replied:
Hm if you insist we can use a decorator .. I'm more used to use classes.  And I'm not a  fan boy "exceptional programming" .. try blocks are expensive and often miss used I think. Compare, less indentations no try block needed:

```python
class SQLiteCursor:
    def __init__(self, db):
        self.database = db
        self.connect = None
        self.cursor = None
    def __enter__(self):
        self.connect = sqlite3.connect(self.database)
        self.connect.row_factory = sqlite3.Row
        self.cursor = self.connect.cursor()
        return self.cursor
    def __exit__(self, exc_type, exc_val, exc_tb):
        if exc_type is None:
            self.connect.commit()
        self.cursor.close()
        self.connect.close()
```

dalf replied:
Scenario:
* some queries crash (whatever the reason)
* cursor and connect are not closed right away but with GC.

The try...finally blocks are here to make the connection whatever happens.

If the cursor is closed with the database it can be:
```python
@contextmanager
def sqlite_cursor(database_url):
	try:
		connect = sqlite3.connect(database_url)
		connect.row_factory = sqlite3.Row
		yield self.connect.cursor()
	finally:
		if connect:
			connect.close()
```

But everything can be fits into the `search` function.

virtadpt replied:
> ```python
>     def __exit__(self, exc_type, exc_val, exc_tb):
>         if exc_type is None:
>             self.connect.commit()
>         self.cursor.close()
>         self.connect.close()
> ```

Searches are read-only operations.  I don't think a SQL COMMIT is required here.

return42 replied:
> Searches are read-only operations. I don't think a SQL COMMIT is required here.

see my https://github.com/searxng/searxng/pull/80#discussion_r636230676

return42 replied:
> Scenario:
>    some queries crash (whatever the reason)

May I do not understand all .. AFAIK for this we use the "with" statement.  Can't speak for the decorator solution, but the `__exit__` method of the class is called,  no matter what was the reason for leaving the "with"" block.

> But everything can be fits into the search function.

Hm .. I do not see the benefit or I misunderstood you.

To get things done: do you like to patch on top what ever solution you prefer .. I can reorder the history afterwards.

dalf commented on searx/engines/sqlite.py:
```diff
@@ -0,0 +1,78 @@
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if exc_type is None:
+            self.connect.commit()
```
There is no need to commit for a SELECT : https://docs.python.org/3/library/sqlite3.html#sqlite3-controlling-transactions

return42 replied:
Not in this use case, but in more general use cases a cursor can execute update statements. I placed the class in the utils section and renamed it:

https://github.com/searxng/searxng/pull/80/files#diff-8d38f1d52f68ec93ec84ebb815fbbb6240fd70d8076dc11ee55021ad58c5658eR617

virtadpt replied:
Searx is a (meta) search engine.  Is there a use case for using it to do anything other than query databases?

return42 replied:
Not at the moment but may be we wish to store some data one day.  This is just a class (context manager) generalized to use fore multi usage, placed in module `utils.py`

dalf commented on searx/engines/sqlite.py:
```diff
@@ -0,0 +1,78 @@
+    def __enter__(self):
+        self.connect = sqlite3.connect(self.database)
+        self.connect.row_factory = sqlite3.Row
+        return self.connect.cursor()
```
The cursor is not closed.

return42 replied:
Normally if the db connection is closed the cursor lands in garbage collector without any cost.  But you are right; explicit closing is better. 

https://github.com/searxng/searxng/pull/80/files#diff-8d38f1d52f68ec93ec84ebb815fbbb6240fd70d8076dc11ee55021ad58c5658eR647

return42 committed:
[pylint] searx/utils.py - activated pylint

activated pylint and fix messages from pylint::

    searx/utils.py:44:0: R0903: Too few public methods (0/2) (too-few-public-methods)
    searx/utils.py:58:0: C0103: Argument name "os" doesn't conform to '(([a-z][a-zA-Z0-9_]{2,30})|(_[a-z0-9_]*))$' pattern (invalid-name)
    searx/utils.py:151:4: R1705: Unnecessary "elif" after "return" (no-else-return)
    searx/utils.py:263:0: C0103: Argument name "d" doesn't conform to '(([a-z][a-zA-Z0-9_]{2,30})|(_[a-z0-9_]*))$' pattern (invalid-name)
    searx/utils.py:322:4: R1705: Unnecessary "else" after "return" (no-else-return)
    searx/utils.py:361:4: R1705: Unnecessary "else" after "return" (no-else-return)
    searx/utils.py:378:12: C0103: Variable name "lc" doesn't conform to '(([a-z][a-zA-Z0-9_]{2,30})|(_[a-z0-9_]*)|([a-z]))$' pattern (invalid-name)
    searx/utils.py:452:0: R1710: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)
    searx/utils.py:464:0: C0103: Argument name "s" doesn't conform to '(([a-z][a-zA-Z0-9_]{2,30})|(_[a-z0-9_]*))$' pattern (invalid-name)
    searx/utils.py:489:4: C0103: Function name "f" doesn't conform to '(([a-z][a-zA-Z0-9_]{2,30})|(_[a-z0-9_]*))$' pattern (invalid-name)
    searx/utils.py:610:7: R1716: Simplify chained comparison between the operands (chained-comparison)

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[enh] utils: implement a context manager for a SQLite cursor

usage in a with statement [2]::

    with SQLiteCursor('test.db') as cur:
        print(cur.execute('select sqlite_version();').fetchall()[0][0])

[1] https://docs.python.org/3/library/stdtypes.html#context-manager-types
[2] https://docs.python.org/3/reference/compound_stmts.html#with

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>
